Fix `FIXME_default_names` by storing static list of 64 none names

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25123 Implement name inference for torch.bmm
* #24474 Implement name inference for torch.dot
* #24471 Implement name inference for addmv, addmv_, mv
* #24469 Implement name inference for expand
* #24306 Implement name inference for mm, addmm, addmm_
* #24921 Fix binary op name inference between unnamed and named tensors.
* #24914 Add helper function Tensor::names()
* #24907 Rename Tensor::names() to Tensor::opt_names()
* **#24885 Fix `FIXME_default_names` by storing static list of 64 none names**
* #24942 Add thread-local-state NamesMode and NoNamesGuard
* #24941 Implement name inference for t(), transpose(...)

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24885 gh/zou3519/109/head

Differential Revision: [D16915803](https://our.internmc.facebook.com/intern/diff/D16915803)

zou3519 committed:
Fix `FIXME_default_names` by storing static list of 64 none names

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

zou3519 committed:
Update on "Fix `FIXME_default_names` by storing static list of 64 none names"

Fix `FIXME_default_names` by storing static list of 64 none names

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24885 gh/zou3519/109/head

nairbv commented on aten/src/ATen/NamedTensor.h:
```diff
@@ -71,9 +71,9 @@ struct CAFFE2_API NoNamesGuard {
-// Everywhere this is used, it is possible to not instantiate the vector by doing
-// some more clever bookkeeping. This is important for performance.
-std::vector<Dimname> FIXME_default_names(size_t len);
+constexpr size_t kMaxNamedTensorDim = 64;
```
it seems like there should be a more generic constant, since 64 max dims is true regardless of named tensors. Maybe we could add a global constant and reference it both in named tensors and in places like WRapDimUtilsMulti.h ?

zou3519 replied:
Sure. I'd need to go audit those places and figure out if there's a way to explain the limitation. I think that that task is outside the scope of this PR; I have an issue open for it here: https://github.com/pytorch/pytorch/issues/23159

zou3519 replied:
@nairbv Aside from having a more generic constant, do you have other comments on this PR? I'd prefer to punt figuring out the (undocumented) dimensionality constraints of a Tensor to later because it isn't very related to this stack of PRs.

nairbv replied:
that was my only concern, yup it's fine to refactor later

zou3519 committed:
Update on "Fix `FIXME_default_names` by storing static list of 64 none names"

Fix `FIXME_default_names` by storing static list of 64 none names

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24885 gh/zou3519/109/head

zou3519 committed:
Update on "Fix `FIXME_default_names` by storing static list of 64 none names"

Fix `FIXME_default_names` by storing static list of 64 none names

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24885 gh/zou3519/109/head

zou3519 committed:
Update on "Fix `FIXME_default_names` by storing static list of 64 none names"

Fix `FIXME_default_names` by storing static list of 64 none names

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24885 gh/zou3519/109/head

zou3519 committed:
Update on "Fix `FIXME_default_names` by storing static list of 64 none names"

Fix `FIXME_default_names` by storing static list of 64 none names

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24885 gh/zou3519/109/head

zou3519 committed:
Update on "Fix `FIXME_default_names` by storing static list of 64 none names"

Fix `FIXME_default_names` by storing static list of 64 none names

Store a static pre-allocated vector of names. When one calls
`default_names`, it gives a const reference to some amount of these
names.

Also make clearer the maximum number of dimensions we support for named
tensors. Right now it is 64 but that number is easy to change. 64
follows some internal pytorch maximum number of dimensions;
TensorIterator reduce ops have a limit of 64 dims.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24885 gh/zou3519/109/head
